### PR TITLE
trie/shadownode: support shadow node hash calculation;

### PR DESCRIPTION
**Pull request by 0xbundler**
### Description

support shadow node hash calculation, opt some trie fields.

### Changes

Notable changes: 
* trie/shadownode: support shadow node hash calculation

### CODE REVIEW

File: trie/trie.go [hunk 1]
```diff
@@ -1026,14 +1050,73 @@ func (t *Trie) resolveShadowNode(epoch types.StateEpoch, cur node, prefix []byte
...
 	}
 }
 
+func (t *Trie) ShadowHash() (*common.Hash, error) {
+	if t.root == nil {
+		return nil, nil
+	}
+	h := newHasher(true)
+	defer returnHasherToPool(h)
+	return t.shadowHash(t.root, h, nil, t.root.getEpoch())
```
Comment (asyukii) [inline trie/trie.go:1067]: should we use `t.root.getEpoch()` or `t.currentEpoch`? or both should be the same?

File: trie/trie.go [hunk 2]
```diff
@@ -1026,14 +1050,73 @@ func (t *Trie) resolveShadowNode(epoch types.StateEpoch, cur node, prefix []byte
...
+		n.shadowNode.ShadowHash = h.shadowNodeHashListToHash(hashList)
+		return h.shadowBranchNodeToHash(&n.shadowNode), nil
+	case valueNode:
+		return nil, nil
+	case hashNode:
+		// resolve temporary, not add to trie
+		rn, err := t.resolveHash(n, prefix)
+		if err != nil {
+			return nil, err
+		}
```
Comment (asyukii) [inline trie/trie.go:1110]: why no need to add the resolved node to trie? 

I'm afraid that the calculation of shadow hash might be quite frequent, so the I/O cost might be higher?

File: trie/trie.go (hunk 1 above)
Comment (0xbundler) [reply trie/trie.go:1067]: when insert/delete a node, it updates the node's epoch and shadowNode epochMap from `t.currentEpoch`, so I prefer to use the node's epoch to represent its epoch.

File: trie/trie.go (hunk 2 above)
Comment (0xbundler) [reply trie/trie.go:1110]: the calculation of shadow hash only once when trie commits, and only fullNode needs resolveHash to calculate its shadowHash, another reason is the later committer will commit all nodes to db, so the temporary node only using in the cache.

The trie node only resolves one time in shadowHash calculation, and is useless in later logic, so I think not adding to trie is ok.

**Review by asyukii — APPROVED** [review]

**Pull request merged**